py/object.go

```diff
@@ -23,24 +23,50 @@ func ObjectRepr(o Object) Object {
 }
 
 // Return whether the object is True or not
-func ObjectIsTrue(o Object) bool {
+func ObjectIsTrue(o Object) (bool, error) {
 	if o == True {
-		return true
+		return true, nil
 	}
 	if o == False {
-		return false
+		return false, nil
 	}
 
 	if o == None {
-		return false
+		return false, nil
 	}
 
 	if I, ok := o.(I__bool__); ok {
 		cmp, err := I.M__bool__()
-		if err == nil && cmp == True {
+		if err != nil {
+			return false, err
+		}
+		if cmp == True {
+			return true, nil
+		} else if cmp == False {
+			return false, nil
+		}
+	}
+	if I, ok := o.(I__len__); ok {
+		l, err := I.M__len__()
+		if err != nil {
+			return false, err
+		}
+		if l.(Int) > 0 {
+			return true, nil
+		} else {
+			return false, nil
+		}
+	}
+	return false, nil
```

Comment:
we should probably try to add a test for this case.

Reply:
`ObjectIsTrue` has been written in a different way, which may look more beautiful, but it may also not fit some checklists,  please review again. At the same time, testing coverage for this method has been added in `filter.py`

Reply:
your rewrite seems to be bug-compliant with the original write up :)

according to the python ref:
"""
[T]he following values are interpreted as false: False, None, numeric zero of all types, and empty strings and containers (including strings, tuples, lists, dictionaries, sets and frozensets). All other values are interpreted as true. 
"""
https://docs.python.org/3/reference/expressions.html#booleans

so with both your rewrite and the previous version, the following script:
```python
class F(object):
 pass

f = F()
if f:
 print("true")

print(all([f]))
```
leads to:
```
true
False
```
whereas with python3 (3.11) this leads to:
```
true
True
```

you may or may not want to fix this as part of this CL.

Reply:
the bug may actually comes from a wrong use of `ObjectIsTrue`: it may be because this function is not meant to be mirroring the python's truthness semantics.

@ncw ? do you confirm/infirm ?

Reply:
This is a simple bug
```c
int
PyObject_IsTrue(PyObject *v)
{
    Py_ssize_t res;
    if (v == Py_True)
        return 1;
    if (v == Py_False)
        return 0;
    if (v == Py_None)
        return 0;
    else if (v->ob_type->tp_as_number != NULL &&
             v->ob_type->tp_as_number->nb_bool != NULL)
        res = (*v->ob_type->tp_as_number->nb_bool)(v);
    else if (v->ob_type->tp_as_mapping != NULL &&
             v->ob_type->tp_as_mapping->mp_length != NULL)
        res = (*v->ob_type->tp_as_mapping->mp_length)(v);
    else if (v->ob_type->tp_as_sequence != NULL &&
             v->ob_type->tp_as_sequence->sq_length != NULL)
        res = (*v->ob_type->tp_as_sequence->sq_length)(v);
    else
        return 1;
    /* if it is negative, it should be either -1 or -2 */
    return (res > 0) ? 1 : Py_SAFE_DOWNCAST(res, Py_ssize_t, int);
}
```
The default return value for CPython is 1, and I only need to modify the default return value to fix it

Reply:
The correct logic should be:

If it is `None` or `False`, return `False`;

If implemented`__bool__` or `__Len__` then return directly, where `__Len__ > 0` is True;

If an error occurs, return `False`;

Otherwise, return `True`;